storage/innobase/fts/fts0fts.cc

```diff
@@ -2607,89 +2607,89 @@ fts_get_next_doc_id(
+           "BEGIN\n"
+           ""
+           "OPEN c;\n"
+           "WHILE 1 = 1 LOOP\n"
+           "  FETCH c INTO my_func();\n"
+           "  IF c % NOTFOUND THEN\n"
+           "    EXIT;\n"
+           "  END IF;\n"
+           "END LOOP;\n"
+           "CLOSE c;");
+
+  *doc_id = 0;
+  error = fts_eval_sql(trx, graph);
+  fts_que_graph_free_check_lock(&fts_table, NULL, graph);
+  return error;
+}
+
+/** This function fetch the Doc ID from CONFIG table, and compare with
 the Doc ID supplied. And store the larger one to the CONFIG table.
+@param table fts  table
+@param cmp_doc_id Doc ID to compare
+@param doc_id     larger document id after comparing "cmp_doc_id" to
+	          the one stored in CONFIG table
+@param trx	  transaction
 @return DB_SUCCESS if OK */
-static MY_ATTRIBUTE((nonnull))
+static
 dberr_t
 fts_cmp_set_sync_doc_id(
-/*====================*/
-	const dict_table_t*	table,		/*!< in: table */
-	doc_id_t		cmp_doc_id,	/*!< in: Doc ID to compare */
-	ibool			read_only,	/*!< in: TRUE if read the
-						synced_doc_id only */
-	doc_id_t*		doc_id)		/*!< out: larger document id
-						after comparing "cmp_doc_id"
-						to the one stored in CONFIG
-						table */
+	const dict_table_t *table,
+	doc_id_t           cmp_doc_id,
+	doc_id_t           *doc_id,
+	trx_t	           *trx=nullptr)
 {
-	trx_t*		trx;
-	pars_info_t*	info;
-	dberr_t		error;
-	fts_table_t	fts_table;
-	que_t*		graph = NULL;
-	fts_cache_t*	cache = table->fts->cache;
-	char		table_name[MAX_FULL_NAME_LEN];
-retry:
-	ut_a(table->fts->doc_col != ULINT_UNDEFINED);
-
-	fts_table.suffix = "CONFIG";
-	fts_table.table_id = table->id;
-	fts_table.type = FTS_COMMON_TABLE;
-	fts_table.table = table;
+	fts_cache_t*	cache= table->fts->cache;
+	dberr_t 	error = DB_SUCCESS;
+	const trx_t*	const created_trx = trx;
 
-	trx = trx_create();
-	if (srv_read_only_mode) {
-		trx_start_internal_read_only(trx);
-	} else {
-		trx_start_internal(trx);
-	}
-
-	trx->op_info = "update the next FTS document id";
-
-	info = pars_info_create();
-
-	pars_info_bind_function(
-		info, "my_func", fts_fetch_store_doc_id, doc_id);
-
-	fts_get_table_name(&fts_table, table_name);
-	pars_info_bind_id(info, "config_table", table_name);
-
-	graph = fts_parse_sql(
-		&fts_table, info,
-		"DECLARE FUNCTION my_func;\n"
-		"DECLARE CURSOR c IS SELECT value FROM $config_table"
-		" WHERE key = 'synced_doc_id' FOR UPDATE;\n"
-		"BEGIN\n"
-		""
-		"OPEN c;\n"
-		"WHILE 1 = 1 LOOP\n"
-		"  FETCH c INTO my_func();\n"
-		"  IF c % NOTFOUND THEN\n"
-		"    EXIT;\n"
-		"  END IF;\n"
-		"END LOOP;\n"
-		"CLOSE c;");
-
-	*doc_id = 0;
-
-	error = fts_eval_sql(trx, graph);
-
-	fts_que_graph_free_check_lock(&fts_table, NULL, graph);
-
-	// FIXME: We need to retry deadlock errors
-	if (error != DB_SUCCESS) {
-		goto func_exit;
+	if (trx == nullptr) {
+		trx = trx_create();
+		if (srv_read_only_mode) {
+			trx_start_internal_read_only(trx);
+		} else {
+			trx_start_internal(trx);
+		}
```

Comment:
If we are creating a transaction here, it would always appear to be a read-only operation, hence we should not call `trx_start_internal(trx)`.